src/compiler/checker.ts

```diff
@@ -15785,6 +15786,23 @@ namespace ts {
             function chooseOverload(candidates: Signature[], relation: Map<RelationComparisonResult>, signatureHelpTrailingComma = false) {
                 candidateForArgumentError = undefined;
                 candidateForTypeArgumentError = undefined;
+                return isSingleNonGenericCandidate ? chooseOverloadSimple(candidates[0], relation, signatureHelpTrailingComma) : chooseOverloadComplex(candidates, relation, signatureHelpTrailingComma);
+            }
+
+            function chooseOverloadSimple(candidate: Signature, relation: Map<RelationComparisonResult>, signatureHelpTrailingComma: boolean): Signature | undefined {

```

Comment:
there's probably no performance impact to declaring a couple of nested functions that don't close over anything in the immediate parent scope, but on the other hand I don't think it reads a whole lot better than

```ts
if (isSingleNonGenericCandidate) {
  // simple code
}
else {
  // original code
}
```

It doesn't help that `chooseOverloadSimple` is an *even* more misleading name than `chooseOverload`. `chooseOverload` should probably be `chooseSignature` or `chooseApplicableSignature` anyway.